File: 1609.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <queue>

struct Solution {
    bool isEvenOddTree(TreeNode* root) {
        queue<TreeNode*> nodes;
        TreeNode *node;
        nodes.push(root);
        int remainder = 1, prevVal = 0, size = 1;
        while (size != 0){
            prevVal = 0;
            for (int i = 0; i < size; ++i){
                node = nodes.front();
                nodes.pop();
                if ((node->val & 1) != remainder){
                    return false;
                }
                if (prevVal != 0){
                    if (remainder == 1 ? prevVal >= node->val : prevVal <= node->val){
                        return false;
                    }
                }
                prevVal = node->val;
                if (node->left != nullptr){
                    nodes.push(node->left);
                }
                if (node->right != nullptr){
                    nodes.push(node->right);
                }
            }
            size = nodes.size();
            remainder ^= 1;
        }
        return true;
    }
};

File: sol/1609.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isEvenOddTree(self, root: Optional[TreeNode]) -> bool:
        nodes = [root]
        remainder = 1
        while len(nodes) != 0:
            prevVal = None
            newNodes = []
            for node in nodes:
                if node.val & 1 != remainder:
                    return False
                if prevVal is not None and (prevVal >= node.val if remainder == 1 else prevVal <= node.val):
                    return False
                prevVal = node.val
                if node.left is not None:
                    newNodes.append(node.left)
                if node.right is not None:
                    newNodes.append(node.right)
            nodes = newNodes
            remainder = 1 - remainder
        return True
